src/librustdoc/html/format.rs

```diff
@@ -485,54 +516,93 @@ impl clean::Path {
                     f.write_str("::")?
                 }
                 if f.alternate() {
-                    write!(f, "{:#}", seg.print(cache, tcx))?;
+                    write!(f, "{:#}", seg.print(cache, tcx, relative_to))?;
                 } else {
-                    write!(f, "{}", seg.print(cache, tcx))?;
+                    write!(f, "{}", seg.print(cache, tcx, relative_to))?;
                 }
             }
             Ok(())
         })
     }
 }
 
-crate fn href(did: DefId, cache: &Cache) -> Option<(String, ItemType, Vec<String>)> {
+crate fn href(
+    did: DefId,
+    cache: &Cache,
+    relative_to: &Vec<String>,
+) -> Option<(String, ItemType, Vec<String>)> {
     if !did.is_local() && !cache.access_levels.is_public(did) && !cache.document_private {
         return None;
     }
 
-    let depth = CURRENT_DEPTH.with(|l| l.get());
-    let (fqp, shortty, mut url) = match cache.paths.get(&did) {
-        Some(&(ref fqp, shortty)) => (fqp, shortty, "../".repeat(depth)),
+    let (fqp, shortty, mut url_parts) = match cache.paths.get(&did) {
+        Some(&(ref fqp, shortty)) => (fqp, shortty, {
+            let fqp = if shortty != ItemType::Module { &fqp[..fqp.len() - 1] } else { &fqp[..] };
+            href_relative_parts(fqp, relative_to)
+        }),
         None => {
             let &(ref fqp, shortty) = cache.external_paths.get(&did)?;
             (
                 fqp,
                 shortty,
                 match cache.extern_locations[&did.krate] {
-                    (.., ExternalLocation::Remote(ref s)) => s.to_string(),
-                    (.., ExternalLocation::Local) => "../".repeat(depth),
+                    (.., ExternalLocation::Remote(ref s)) => {
+                        let s = if s.ends_with('/') { &s[..s.len() - 1] } else { &s[..] };
+                        let fqp = if shortty != ItemType::Module {
+                            &fqp[..fqp.len() - 1]
+                        } else {
+                            &fqp[..]
+                        };
+                        let mut s = vec![&s[..]];
+                        s.extend(fqp[..].iter().map(String::as_str));
+                        s
+                    }
+                    (.., ExternalLocation::Local) => {
+                        let fqp = if shortty != ItemType::Module {
+                            &fqp[..fqp.len() - 1]
+                        } else {
+                            &fqp[..]
+                        };
```

Comment:
You use this a lot; maybe assign it to a local variable at the top of the function?

Reply:
Sure, let's do that.